From: velsinki <112010-velsinki@users.noreply.gitlab.gnome.org>
Date: Thu, 16 Mar 2023 22:35:11 +0100
Subject: sound: Prevent duplicate sound device entries

For unknown reasons, GVC mixer control can sometimes signal a new
device with the same id as one that was added before. This means that
in `device_added_cb`, a duplicate entry with that id is created, in my
case with a different name. However, the last one added is valid, but
that one cannot be selected because all other logic in the sound panel
assumes the first hit in `get_iter` is valid. This breaks sound input
selection then.

The fix is easy; only add a new list entry if none with that id exists.

Origin: https://gitlab.gnome.org/GNOME/gnome-control-center/-/commit/eacef5a
---
 panels/sound/cc-device-combo-box.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/panels/sound/cc-device-combo-box.c b/panels/sound/cc-device-combo-box.c
index 5db60a0..b0f5d47 100644
--- a/panels/sound/cc-device-combo-box.c
+++ b/panels/sound/cc-device-combo-box.c
@@ -33,6 +33,8 @@ struct _CcDeviceComboBox
 
 G_DEFINE_TYPE (CcDeviceComboBox, cc_device_combo_box, GTK_TYPE_COMBO_BOX)
 
+static gboolean get_iter (CcDeviceComboBox *self, guint id, GtkTreeIter *iter);
+
 static void
 device_added_cb (CcDeviceComboBox *self,
                  guint             id)
@@ -65,7 +67,9 @@ device_added_cb (CcDeviceComboBox *self,
   if (gvc_mixer_ui_device_get_icon_name (device) != NULL)
     icon_name = g_strdup_printf ("%s-symbolic", gvc_mixer_ui_device_get_icon_name (device));
 
-  gtk_list_store_append (self->device_model, &iter);
+  if (!get_iter (self, id, &iter))
+    gtk_list_store_append (self->device_model, &iter);
+
   gtk_list_store_set (self->device_model, &iter,
                       0, label,
                       1, icon_name,
